lib/src/command/unpack.dart

```diff
@@ -0,0 +1,220 @@
+  @override
+  String get name => 'unpack';
+
+  @override
+  String get description => '''
+Downloads a package and unpacks it in place.
+
+For example:
+
+  $topLevelProgram pub unpack foo
+
+Downloads and extracts the lastest stable package:foo from pub.dev in a
+directory `foo-<version>`.
+
+  $topLevelProgram pub unpack foo:1.2.3-pre
+
+Downloads and extracts package:foo version 1.2.3-pre in a directory
+`foo-1.2.3-pre`.
+
+  $topLevelProgram pub unpack foo --output=archives
+
+Downloads and extracts latest stable version of package:foo in a directory
+`archives/foo-<version>`.
+
+  $topLevelProgram pub unpack 'foo:{hosted:"https://my_repo.org"}'
+
+Downloads and extracts latest stable version of package:foo from my_repo.org
+in a directory `foo-<version>`.
+
+
+Will resolve dependencies in the folder unless `--no-resolve` is passed.
+''';
+
+  @override
+  String get argumentsDescription => 'package-name[:constraint]';
+
+  @override
+  String get docUrl => 'https://dart.dev/tools/pub/cmd/pub-unpack';
+
+  @override
+  bool get takesArguments => true;
+
+  UnpackCommand() {
+    argParser.addFlag(
+      'resolve',
+      help: 'Whether to do pub get in the downloaded folder',
+      defaultsTo: true,
+      hide: log.verbosity != log.Verbosity.all,
+    );
+    argParser.addOption(
+      'output',
+      abbr: 'o',
+      help: 'Download and extract the package in this dir',
+      defaultsTo: '.',
+    );
+  }
+
+  static final _argRegExp = RegExp(
+    r'^(?<name>[a-zA-Z0-9_.]+)'
+    r'(?::(?<descriptor>.*))?$',
+  );
+
+  @override
+  Future<void> runProtected() async {
+    if (argResults.rest.isEmpty) {
+      usageException('Provide a package name');
+    }
+    if (argResults.rest.length > 1) {
+      usageException('Please provide only a single package name');
+    }
+    final arg = argResults.rest[0];
+    final match = _argRegExp.firstMatch(arg);
+    if (match == null) {
+      usageException('Use the form package:constraint to specify the package.');
+    }
+    final parseResult = _parseDescriptor(
+      match.namedGroup('name')!,
+      match.namedGroup('descriptor'),
+    );
+
+    if (parseResult.ref.description is! HostedDescription) {
+      fail('Can only fetch hosted packages.');
+    }
+    final versions = await parseResult.ref.source
+        .doGetVersions(parseResult.ref, null, cache);
+    final constraint = parseResult.constraint;
+    if (constraint != null) {
+      versions.removeWhere((id) => !constraint.allows(id.version));
+    }
+    if (versions.isEmpty) {
+      fail('No matching versions of ${parseResult.ref.name}.');
+    }
+    versions.sort((id1, id2) => id1.version.compareTo(id2.version));
+
+    final id = versions.last;
+    final name = id.name;
+
+    final outputArg = argResults['output'] as String;
+    final destinationDir = p.join(outputArg, '$name-${id.version}');
+    if (entryExists(destinationDir)) {
+      fail('Target directory `$destinationDir` already exists.');
+    }
```

Comment:
Maybe it would be nice with an option to override the existing folder?

Reply:
Or allow a copy? This is a nit.

Reply:
I made a `--force` flag to overwrite the folder if it exists